In [1]:
# Initialize Otter
import otter
grader = otter.Notebook("hw1.ipynb")

# Hw 1: Data Science Workflow and More NumPy 📈

### Collaborators and Sources
In addition to recording your **collaborators** (TAs, peers, group members, roommates, etc.) on this homework in the cell below, you are required to **cite/indicate all external sources** used when finishing this assignment. 
External sources are defined as anything that is not considered course material, such as online sources (webpages, blog posts, etc.), content generated by AI systems, books, etc.

When using external sources, indicate what kind of external sources (e.g. stack overflow, WashU chatGPT, etc.) you used in the cell below and then provide more specific citations (such as links to webpages or in case of AI generated asnwers the actual sources (links provided by the AI system) and the prompt used to generate the answers) with your answer to **each specific problem**.  


Note that these citations will not free you from your obligation to submit your _own_ code and write-ups, however, they will be taken into account during the grading and regrading process, **especially** when two or more submissions closely resemble each other. Working with each other is ok, as long as you cite who you worked with and you don't copy anyone's answers directly!

### Submission instructions
* Submit this Python notebook on Gradescope.
* Execute all cells and save your notebook prior to submission. 
* Don't forget to commit and push to your Github repo!
* **Do not change the number of cells!** Your submission notebook should have exactly one code cell per problem. 
* Do **not** remove the `# your code here` line. Add your solution **after** that line. 

## Let's get started!

For this homework, we will be gaining more exposure to NumPy and exploring the data science workflow. We will also practice plotting data.  **Important:** Ensure that all plots have the necessary components that a plot should have: axes labels and a title (and a legend when it is appropriate). You must have all components of a plot on every plot for the entire semester!

Work your way through these problems. They will also provide some more practice for working with NumPy arrays. 
> Use NumPy arrays and functions in **all** problems unless explicitly stated otherwise! 

Frequently **save** your notebook!

## 1. Churn Analysis: Data Science Workflow

Consider the following data science use case:

> *Assume you just landed a great analytical job with MegaTelCo, one of the largest telecommunication firms in the United States. They are having a major problem with customer retention in their wireless business. In the mid-Atlantic region, 20% of cell phone customers leave when their contracts expire, and it is getting increasingly difficult to acquire new customers. Since the cell phone market is now saturated, the huge growth in the wireless market has tapered off. Communications companies are now engaged in battles to attract each other’s customers while retaining their own. Customers switching from one company to another is called churn, and it is expensive all around: one company must spend on incentives to attract a customer while another company loses revenue when the customer departs.*
>
> *You have been called in to help understand the problem and to devise a solution. Attracting new customers is much more expensive than retaining existing ones, so a good deal of marketing budget is allocated to prevent churn. Marketing has already designed a special retention offer. Your task is to devise a precise, step-by-step plan for how the data science team should use MegaTelCo’s vast data resources to decide which customers should be offered the special retention deal prior to the expiration of their contracts.*

<!-- BEGIN QUESTION -->


### Problem 1.1

**Write-up!** Is this a scientific, social, or business problem? Briefly justify your answer.

_This is a business problem because we are dealing with helping a company. The company wants to retain more customers than their retention already so their main objective is deciding customer retention. We are working directly with a business on how to come up with a solution to keep their customers, therefore, keeping their profits._

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->

### Problem 1.2

**Write-up!** Using the **data science workflow** discussed in class, describe how you would takle this task. Make sure to list every step and briefly describe the step _in the context of this specific use case_. 

> **Hint**: Include ideas on what data you might use and how they would be used. 

_Scientific, social or business problem: We have already decided this is a business problem, mostly. 

Data problem: We identitied this problem as a business problem so we can create a data question. For example, can we predict which customers are most likely to leave when their contrasts expire (so they can be offered the special retention deal)?

Collect data: We will need to collect data like MegaTelCo's vast data resources. For example, service usage like texts and messages. There is also demographic data like income range, age, and their locations. Contract details (length of contract, the type of contract they're under, when they'll have to renew the contract), their billing/payment histories, and if they have previous history of churning (jumping from one company to another).

Clean and format data: We'll have to fix and clean up the data before we can start analyzing it. We'll have to deal with any missing values (like incomplete info on a customer's profile), correct errors and typos, remove duplicates, etc. 

Explore and understand data: We can now use data to look for patterns. For example, does the different contrast types affect retention? Are there certain behaviors that customers with churning history exhibit that customers without churning history do not? Are there certain usage levels that are at higher risk? 

Use data to create solution: We can use our datas to predict high-risk customers before their contrasts with MegaTelCo ends. Also, we can decide which ones can be offered the special retention deal from our data. 

Communicate results: We can share this data to people by making graphs of our findings, create recommendations, and estimations on how many customers will churn as well as how much revenue can be saved. 
_

<!-- END QUESTION -->

## 2. Practice NumPy Indexing

Indexing, especially with NumPy, can be a tricky feature to truly wrap one's head around, but the benefits of (working towards) mastering it make it a worthy endeavor. The more practice you get, the easier it will become — eventually you won't be able to even imagine how else you would do things.

In this section, we will start with some review and then move on to more complex features.

### Basic Indexing
At the risk of belaboring this topic, let's quickly do some practicing. The following cell produces `yet_another_array` using the `np.random.randint` function, which we will use in the next few problems.

In [2]:
import numpy as np
yet_another_array = np.random.randint(11, size=10)
yet_another_array

array([ 4,  4,  2,  8,  0,  4, 10,  8,  4,  2])

### Problem 2.1

**Do this!** Retrieve the 4th element of `yet_another_array`.

*Note: this question has hidden tests, or is graded on style of code and not just answer alone.*

In [3]:
element = yet_another_array[3]

### Problem 2.2

Here's something new: specifically for NumPy arrays, you can also pass in a Python list or NumPy array of indicies to retrieve.

For example, `some_array[[0, 2, 4, 6]]`.


**Do this!** Retrieve the 2nd, 5th, and 9th elements of `yet_another_array`.

*Note: this question has hidden tests, or is graded on style of code and not just answer alone.*

In [4]:
elements = yet_another_array[[1, 4, 8]]

In [5]:
grader.check("q2b")

q2b results: All test cases passed!

### Problem 2.3

Let's do this with a 2D array, too. Here we will generate `a_2d_array` using the `np.random.rand` function.

*Note: this question has hidden tests, or is graded on style of code and not just answer alone.*

In [6]:
a_2d_array = np.random.rand(5, 5)
a_2d_array

array([[0.36502612, 0.78610937, 0.32733902, 0.86715683, 0.39705407],
       [0.74678264, 0.5103292 , 0.12011909, 0.58983911, 0.33814086],
       [0.40758051, 0.6025015 , 0.91999564, 0.94464696, 0.87469629],
       [0.90187198, 0.93400139, 0.88162146, 0.43693729, 0.14851081],
       [0.9216223 , 0.42602053, 0.97310243, 0.98319459, 0.16346995]])

Remember that you can index into a 2D array like this: `another_2D_array[row, column]`, where `row` and `column` are indicies, slices, `:`s, or some mix of these three.

**Do this!** Retrieve the value at index `(3, 4)` from `a_2d_array`.

In [7]:
a_2d_element = a_2d_array[3,4]

### Slice Indexing

You've already gotten familiar with slices, but here is some more practice.

### Problem 2.4

**Do this!** Retrieve the 6th, 7th, and 8th elements of `yet_another_array`.

*Note: this question has hidden tests, or is graded on style of code and not just answer alone.*

In [8]:
more_elements = yet_another_array[5:8]

In [9]:
grader.check("q2d")

q2d results: All test cases passed!

### Problem 2.5

Here is a reminder of what `a_2d_array` looks like.

*Note: this question has hidden tests, or is graded on style of code and not just answer alone.*

In [10]:
a_2d_array

array([[0.36502612, 0.78610937, 0.32733902, 0.86715683, 0.39705407],
       [0.74678264, 0.5103292 , 0.12011909, 0.58983911, 0.33814086],
       [0.40758051, 0.6025015 , 0.91999564, 0.94464696, 0.87469629],
       [0.90187198, 0.93400139, 0.88162146, 0.43693729, 0.14851081],
       [0.9216223 , 0.42602053, 0.97310243, 0.98319459, 0.16346995]])

**Do this!** Retrieve the values in the 3rd column of `a_2d_array`.

In [11]:
column = a_2d_array[:, 2]

In [12]:
grader.check("q2e")

q2e results: All test cases passed!

### Logical Indexing

Now for something complete new! You might be interested in finding all of the values in an array that fulfills some condition like _get all values greater than 5_. 

> [🐍 **Python Feature** 🐍]: NumPy enables this by supporting [**Boolean (or logical) indexing**](https://docs.scipy.org/doc/numpy-1.13.0/user/basics.indexing.html#boolean-or-mask-index-arrays).

This idea is also referred to as **masking**. You can use an array of boolean values _of the same shape_ as the target array as the "index" into the target array. This will return all of the values that are in the same position as the `True` values from the boolean arrays. These logical arrays are called "masks" because they are analogous to masks, which let some of the underlying surface show through, but hide the rest.

Let's try making our first logical array. In the cell below, we use a comparison operation on `yet_another_array`, returning an array of the same size with `True` in the positions where the values meet the condition.

In [13]:
yet_another_array > 5

array([False, False, False,  True, False, False,  True,  True, False,
       False])

Now let's see how this array "lines up" with `yet_another_array`.

In [14]:
yet_another_array

array([ 4,  4,  2,  8,  0,  4, 10,  8,  4,  2])

Here is another example using an array that is not sorted already.

In [15]:
shuffled_array = np.random.permutation(yet_another_array)
shuffled_array

array([ 8,  0,  2,  4,  4,  4,  4,  2, 10,  8])

In [16]:
shuffled_array > 5

array([ True, False, False, False, False, False, False, False,  True,
        True])

Is this what you would expect? Let's actually get those values.

In [17]:
shuffled_array[shuffled_array > 5]

array([ 8, 10,  8])

### Problem 2.6

**Do this!** Retrieve the elements of `yet_another_array` that are even (**Hint**: `%`).

*Note: this question has hidden tests, or is graded on style of code and not just answer alone.*

In [18]:
even_array = yet_another_array[yet_another_array % 2 == 0] 

## 3. Model Evaluation and More Numpy


### Error Measure

Lets assume we have the following predictions `y_pred` and observed targets `y_true` for evaluation. 

In [19]:
y_hat = np.array([0, 2, 1, 2, 1, 0, 0, 2, 1, 0, 1, 0, 2, 1, 2, 1, 1, 
                   0, 2, 0, 0, 1, 0, 2, 1, 1, 0, 1, 1, 0, 1, 2, 2, 0, 
                   0, 0, 2, 1, 0, 0, 2, 2, 0, 1, 1])
y_test = np.array([0, 2, 2, 2, 1, 0, 0, 1, 2, 0, 1, 0, 2, 1, 2, 1, 1,
                   0, 2, 0, 0, 1, 0, 2, 1, 1, 0, 1, 1, 0, 1, 2, 2, 0, 
                   0, 0, 2, 1, 0, 0, 2, 2, 0, 1, 1])

### Problem 3.1

**Do this!** Complete the following function so that given two arrays of labels, `y_test` and `y_hat`, it will return the average classification error, `avg_err`. Use **Boolean operations** on the arrays instead of a for loop.

*Note: this question has hidden tests, or is graded on style of code and not just answer alone.*

In [20]:
def avg_error(y_true, y_pred):
    '''
    Computes the classification error given Y_TRUE and Y_PRED
    as a value from 0 to 1
    '''


    
    avg_err = np.mean(y_true != y_pred)
    
    return avg_err

In [21]:
err_score = avg_error(y_hat, y_test)
print(f'Error: {100*err_score:0.2f}%')

Error: 6.67%


<!-- BEGIN QUESTION -->

**Write-up!** How does the _average classification error_ computed above relate to the _average classification accuracy_ that you computed in Lab 1? State a mathematical formula of how you can compute the average error from the average accuracy. 

_The average error is used to measure the amount of incorrect predictions whilst the average accuracy is used to measure the amount of correct predictions. So they are intrinsicallly related to each other and if you add them together, they should equal 1 (a whole). So if you do == you get the average accuracy, if you do != it's the average error. A formula you could use is avg_error + avg_accuracy = 1. To get the average error above, the formula would look like: avg_error = 1 - avg_accuracy._

<!-- END QUESTION -->

### Evaluating a Model using all the Features

As you might have concluded in your communication your findings step, one way to improve our predictions could be to use all the features instead of just one. Lets do that now and compare the results. We saved the processed data for you. This is the same data that you created and used in Lab1. 

In [22]:
data = np.load('utility/data/processed_iris_data.npy')
data[:5,:]

array([[5.1, 3.5, 1.4, 0.2, 0. ],
       [4.9, 3. , 1.4, 0.2, 0. ],
       [4.7, 3.2, 1.3, 0.2, 0. ],
       [4.6, 3.1, 1.5, 0.2, 0. ],
       [5. , 3.6, 1.4, 0.2, 0. ]])

### Problem 3.2

**Do this!** Create `X` and `y` arrays. Now with X including all four features. Then reproduce the steps from Lab1 (D. Building and evaluating a model and making predictions). 

*Note: this question has hidden tests, or is graded on style of code and not just answer alone.*

In [23]:
X, y = data[:,:-1], data[:,-1]

...

Ellipsis

### Problem 3.3

**Do this!** Make the training and testing sets exactly with the same configuration as in Lab1. 
> Note: This is **very important** since it will let us compare the new results with the ones we got in the lab (since we are using the _same_ training and test splits).

*Note: this question has hidden tests, or is graded on style of code and not just answer alone.*

In [24]:
from sklearn.model_selection import train_test_split

# train_test_split will be used a lot in this class, be sure to remember how to use it!
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=217)

#Collaborators: Valeria Garcia-Hernandez. Shared her lab code with me because I was one of the teams that could not finish on time due to issues with the graph parts for Lab 1. 
...

Ellipsis

### Problem 3.4

**Do this!** Train the Model on the training data that you created in the previous step. 

*Note: this question has hidden tests, or is graded on style of code and not just answer alone.*

In [25]:
from utility.util import SimpleModel
model = SimpleModel()

In [26]:
model.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=1)

In [27]:
grader.check("q3d")

q3d results: All test cases passed!

### Problem 3.5

**Do this!** Make predictions on the test set and compare them to the ground truth by computing the average classification error. 
> **Hint:** use your `avg_error()` function from above. 

*Note: this question has hidden tests, or is graded on style of code and not just answer alone.*

In [28]:
predictions = model.predict(X_test)

score = avg_error(y_test, predictions)

print(f'Avg Error: {100*score:0.2f}%')

Avg Error: 4.44%


In [29]:
grader.check("q3e")

q3e results: All test cases passed!

<!-- BEGIN QUESTION -->

### Problem 3.6

**Write-up!** Report your average classification error from both the lab and this homework. Did the the model in this homework perform better or worse compared to the lab? Why do you think the performace imporved or did not imporve?    

_The average classification error from this homework was 4.44%. The accuracy for Lab1 would've been 91.11%, making the average classification error 8.89%. The model from this homework performed better compared to the lab and I think the performance improved because in 3.2, it said "Create `X` and `y` arrays. Now with X including all four features." This makes me to believe by including all features, the model is able to capture the data and the data patterns better, decreasing error and incrdasing accuracy._

<!-- END QUESTION -->

**That's it! Follow the submission instructions below:**
* Restart your kernel and run all the cells
* Save your notebook
* Upload the .ipynb notebook directly to Gradescope